spec/requests/api/v2/providers/courses/update_sites_spec.rb

```diff
@@ -0,0 +1,77 @@
+require "rails_helper"
+
+describe 'PATCH /providers/:provider_code/courses/:course_code with sites' do
+  let(:organisation) { create :organisation }
+  let(:provider)     { create :provider, organisations: [organisation] }
+  let(:user)         { create :user, organisations: [organisation] }
+  let(:payload)      { { email: user.email } }
+  let(:token)        { build_jwt :apiv2, payload: payload }
+  let(:credentials) do
+    ActionController::HttpAuthentication::Token.encode_credentials(token)
+  end
+
+  let(:course) { create :course, provider: provider }
+  let(:site_to_add) { create :site, provider: provider }
+  let(:unwanted_site) { create :site, provider: provider }
+  let(:existing_site) { create :site, provider: provider }
+
+  before do
+    course.add_site!(site: existing_site)
+    course.add_site!(site: unwanted_site)
+    perform_request
+  end
+
+  let(:jsonapi_data) do
+    {
+      "_jsonapi" => {
+        "data" => {
+          "id" => course.id.to_s,
+          "type" => "courses",
+          "relationships" => {
+            "sites" => {
+              "data" => [
+                { "type" => "sites", "id" => existing_site.id.to_s },
+                { "type" => "sites", "id" => site_to_add.id.to_s }
+              ]
+            }
+          }
+        }
+      }
+    }
+  end
+
+  def perform_request
+    patch "/api/v2/providers/#{course.provider.provider_code}" \
+            "/courses/#{course.course_code}",
+          headers: { 'HTTP_AUTHORIZATION' => credentials },
+          params: jsonapi_data
+  end
+
+  context 'course has some sites' do
+    it 'returns http success' do
+      expect(response).to have_http_status(:success)
+    end
+
+    it "adds a new site" do
+      expect(course.reload.sites.exists?(site_to_add.id)).to be(true)
+    end
+
+    it "leaves existing site in place" do
+      expect(course.reload.sites.exists?(existing_site.id)).to be(true)
+    end
+
+    it "removes an unwanted site" do
+      expect(course.reload.sites.exists?(unwanted_site.id)).to be(false)
+    end
+
+    context "course is running" do
+      before do
+        course.publish_sites
+      end
+
+      it "suspends an unwanted site" do
+        expect(course.reload.sites.exists?(unwanted_site.id)).to be(false)
```

Comment:
On second thoughts @tvararu perhaps we do want to dive into `site_statuses` here and check it's suspended? Now that everything's working I'm not sure I'm happy with the two "unwanted site" tests here, they're a bit surprising/magical.

Reply:
Thanks, that uncovered a bug, fixed